src/typed/typecore.ml

```diff
@@ -509,6 +529,17 @@ and type_expect_ ?in_function ?(recarg=Rejected) env sexp ty_expected_explained
       exp_type = ifso.exp_type;
       exp_env = env
     }
+  | PExpWhile(scond, sbody) ->
+    let cond = type_expect env scond (mk_expected ~explanation:While_loop_conditional Builtin_types.type_bool) in
+    let body = type_expect env sbody ty_expected_explained in
+    re {
+      exp_desc = TExpWhile(cond, body);
+      exp_loc = loc;
+      exp_extra = [];
+      (* While loops don't evaluate to anything *)
```

Comment:
I'm not a huge fan of this, given our `if`/`else` semantics, but I don't have a good alternative. I'd argue that `while` loops should evaluate to something, but it's hard to say what they should be in the case of `while (false) { ... }`...

Reply:
I thought the same thing and wanted them to evaluate to something too. I think I disagree about it not being similar to `if`/`else`, since it's the same deal as having a one-sided `if`. A couple of options, though:

1) We could also/otherwise have `do {} while` loops, which would evaluate to something:
```grain
let i = box(5);
do {
  i := unbox(i) - 1;
} while (unbox(i) > 0)
# Evaluates to box(0)
```

2) Some languages have this kinda thing:
```grain
while (someBool) {
  # Do some stuff
} else {
  # Some other stuff
}
```

I think I'd prefer option 1, since I also don't really like one-sided `if`s (which I don't know if we even have them implemented right now, though we do have a parsing rule for them) because the semantics change based on whether or not you've got that `else` clause (which I think people could possibly get behind, especially if we had a helpful message when they did it, but I feel like I rather avoid it if we could, but I'll think on it more). So instead of something like one-sided `if`, I might be more in favor of calling it `when` or something similar like other languages do (but also maybe something different, because I feel like it's confusing when you also have async built into your language).